Integrate Neighboring Road Names Retrieval

pushpak-gote-mapup opened the pull request:
### **User description**
This script gets all neighbouring road names from the state dataset for bridge points.


___

### **PR Type**
Enhancement, Other


___

### **Description**
- Added a new script `get_neighbouring_roads.py` to retrieve and process neighboring road names from state datasets.
- Implemented functions to read GeoPackage files, transform CRS, create buffers, perform spatial joins, and group/aggregate data.
- Included error handling and logging for better traceability and debugging.
- Saved intermediate and final results to GeoPackage and CSV files for further analysis.


___



### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>get_neighbouring_roads.py</strong><dd><code>Script for retrieving and processing neighboring road names</code></dd></summary>
<hr>

mile-point-approach/get_neighbouring_roads.py

<li>Added script to read, transform, and process geospatial data.<br> <li> Implemented buffer creation, spatial joins, and data aggregation.<br> <li> Included error handling and logging for geoprocessing steps.<br> <li> Saved intermediate and final results to GeoPackage and CSV files.<br>


</details>


  </td>
  <td><a href="https://github.com/mapup/edit-osm-add-missing-bridge-for-truck-restriction/pull/32/files#diff-fba8e0bdc0f585e389c50d32cbc53b53638cf38e3464ba93956c928c9ec56e65">+246/-0</a>&nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

maneesh-mapup commented on mile-point-approach/get_neighbouring_roads.py:
```diff
@@ -0,0 +1,246 @@
+    """Enum for output control flags."""
+    SAVE_INTERMEDIATE_GEOPACKAGES = True
+
+class GeoprocessingError(Exception):
```
@pushpak-gote-mapup Respond to AI code review before merging

pushpak-gote-mapup replied:
Yes @maneesh-mapup, that's why I haven't added any reviewers yet. Will be modifying the code as per suggestions from the AI agent first.

pushpak-gote-mapup commented:
Solved PR Agent issues:
[Error Handling](https://github.com/mapup/edit-osm-add-missing-bridge-for-truck-restriction/pull/32/files#diff-fba8e0bdc0f585e389c50d32cbc53b53638cf38e3464ba93956c928c9ec56e65R37-R37)
The custom exception GeoprocessingError is raised in multiple functions but there is no specific handling logic for different types of geoprocessing errors which might be helpful for debugging specific issues.
**Solved:**
Created more specific custom exceptions (FileReadError, CRSTransformError, BufferCreationError, SpatialJoinError, and GroupingError) allowing for more granular error handling.
<img width="612" alt="Screenshot 2024-08-01 at 12 31 48 PM" src="https://github.com/user-attachments/assets/69ba5857-4795-4ed0-adec-027997d3bf69">

[Performance Concern](https://github.com/mapup/edit-osm-add-missing-bridge-for-truck-restriction/pull/32/files#diff-fba8e0bdc0f585e389c50d32cbc53b53638cf38e3464ba93956c928c9ec56e65R104-R126)
The method perform_spatial_join uses a list comprehension inside a try block to perform multiple spatial joins, which might be inefficient and error-prone if the datasets are large. Consider optimizing this by reducing redundancy and improving error handling.
**Solved:**
The perform_spatial_join function has been optimized to handle errors for individual predicates without failing the entire operation. We now use a dictionary to store results for each predicate, allowing us to continue even if one join fails.
<img width="673" alt="Screenshot 2024-08-01 at 12 34 43 PM" src="https://github.com/user-attachments/assets/42a0905c-43bf-4962-9d58-c3a129695d0d">


[Code Clarity](https://github.com/mapup/edit-osm-add-missing-bridge-for-truck-restriction/pull/32/files#diff-fba8e0bdc0f585e389c50d32cbc53b53638cf38e3464ba93956c928c9ec56e65R16-R25)
The use of Enums for file paths and buffer distances might be over-engineering for configurations that could be simpler as constants or configuration files, especially if these values are not extensively reused or changed.
**Solved:**
No changes need to be made

Have also implemented following PR Code Suggestions:
  - Improve CRS comparison to ensure robustness and compatibility
  - Use a context manager for file operations to enhance reliability
  - Add error handling to the grouping operation to enhance data processing reliability

pushpak-gote-mapup commented:
Here's the link to the document explaining logic:
https://docs.google.com/document/d/1nbfJcDjXb9UPJxEbIomd1MRPvL3snTu6DBuBAuMX_Dc/edit?usp=sharing

pushpak-gote-mapup commented:
Hello @maneesh-mapup and @Abeer-Mapup, I have made the changes as per the PR agent, please review the code.

Abeer-Mapup commented:
@pushpak-gote-mapup here are some suggestions:
The buffer creation in the create_buffer function should account for the units of the CRS. For example, if the CRS is in meters, the buffer distance should be in meters. Ensure the buffer distances provided in the BufferDistance enum are in the correct units for the CRS being used.

The group_and_aggregate function assumes certain column names exist. If these columns are missing, it will raise a KeyError. Consider adding a check for required columns before performing the grouping operation

Type hinting: While the script uses type hints in some places, it could benefit from more consistent use throughout, especially for function return types.

pushpak-gote-mapup commented:
@Abeer-Mapup I have modified the create_buffer function to account for CRS units and added the remaining type hints. Additionally, the KeyError in the group_and_aggregate function is already being handled.

varun-andhra-mapup commented:
Hey @pushpak-gote-mapup, here's the review:

### Strengths:
 - Modularity: Clear separation of tasks with well-defined functions and use of Enums for configuration.
 - Security: Proper error logging and custom exceptions enhance debugging and error handling.
 - Performance: Efficient use of geopandas and clear handling of CRS transformations and buffering.

### Improvements:
 - Modularity: Further break down the main function into smaller functions for better readability.
 - Security: Validate file paths and user inputs to prevent potential issues.
 - Performance: Optimize buffer creation and consider memory management for large datasets.

pushpak-gote-mapup commented:
Hello @varun-andhra-mapup, I have made the suggested improvements. 
> Hey @pushpak-gote-mapup, here's the review:
> 
> ### Strengths:
> * Modularity: Clear separation of tasks with well-defined functions and use of Enums for configuration.
> * Security: Proper error logging and custom exceptions enhance debugging and error handling.
> * Performance: Efficient use of geopandas and clear handling of CRS transformations and buffering.
> 
> ### Improvements:
> * Modularity: Further break down the main function into smaller functions for better readability.
> * Security: Validate file paths and user inputs to prevent potential issues.

Abeer-Mapup commented on mile-point-approach/get_neighbouring_roads.py:
```diff
@@ -0,0 +1,393 @@
+        ValueError: If the CRS units are not supported.
+    """
+    try:
+        crs = gdf.crs
```
My feedback

Abeer-Mapup commented on mile-point-approach/get_neighbouring_roads.py:
```diff
@@ -0,0 +1,393 @@
+        ValueError: If the CRS units are not supported.
+    """
+    try:
+        crs = gdf.crs
```
my feedback

pushpak-gote-mapup replied:
Too care of it

Abeer-Mapup commented on mile-point-approach/get_neighbouring_roads.py:
```diff
@@ -0,0 +1,393 @@
+        elif crs.is_geographic:
+            # For geographic CRS, convert degrees to meters (approximate)
+            distance = buffer_distance.value / 111000  # 1 degree ≈ 111 km
+        else:
```
another feedback

pushpak-gote-mapup replied:
Finished it

maneesh-mapup reviewed (approved):
Good to go
